src/core/mac/mac.cpp

```diff
@@ -553,7 +553,7 @@ void Mac::UpdateIdleMode(void)
 #if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
         if (IsCslEnabled())
         {
-            mLinks.CslSample(mRadioChannel);
```

Comment:
I think there would be some problems with this change.

The idea of this code block in `UpdateIdleMode` is: we prioritize other MAC layer operations than 'CslSample' (for example, TX) and only start CslSample when MAC enters Idle mode. A common case is, when a TX happens, the CSLSample state should stop. Be aware that we must ensure the MAC layer only be in one certain state and the SubMac layer should only do one operation at one time.

With current change, the state in SubMac layer will not fully submit to the Mac layer. Because the CSL timer in SubMac will also directly control the radio. This will cause problems like: the MAC layer starts a TX operation. But before SubMac layer completes it, the CSL timer fires. Then it will probably sleep, which breaks the TX operation. And that also breaks the 'rule' of Mac layer. (It should only do one operation at one time and the SubMac layer should be fully consistent with it)

I'm not sure I described it well enough. IIRC, when we implemented CSL at first, I had a discussion with @abtink. Based on this, I think we cannot have this change.

@abtink Could you help confirm?

Reply:
Maybe the `kStateCslSample` can be kept in the SubMac for that purpose? While still handling the start/stop of the CSL timer from the MAC and adding some checks for the transitions (i.e., only enter CslSample state from Idle).

Reply:
I agree with @Irving-cl . I think removing `kStateCslSample` from `SubMac` may not be a good idea. Is there maybe a benefit to remove this state?

The idea behind current model is that `SubMac` acts like a "more capable" radio. It can be in a single state at a time and it provides a bunch of APIs to next layers (`Mac`) to control it and put it in different states: 
- `Send()` to enter tx
- `Receive()` to enter rx mode
- `Sleep()` to go to sleep mode
- We also have a fancier `CslSample()` which puts `SubMc` in CSL sampling state (which it then does CSL sampling based on period/channel/etc).